File: web/src/api/snapshot.ts
/**
 * 虚拟机快照（F-2-07）。
 *
 * 创建与恢复都是耗时操作（要复制或回滚整个磁盘镜像），因此接口只返回任务
 * 标识，进度在任务中心与列表里的状态字段上体现。
 */
import { del, get, post } from './client'
import type { TaskRef } from './vm'

/**
 * 快照种类。
 *
 * 这是从「是否含内存」与「创建时的运行态」**推导**出来的，不是用户选的：
 * 它们是虚拟化层的实现细节，用户关心的是「要不要保存运行现场」。
 */
export type SnapshotKind = 'internal' | 'external'

export type SnapshotStatus = 'creating' | 'ready' | 'restoring' | 'deleting' | 'error'

export interface Snapshot {
  id: number
  name: string
  description?: string
  kind: SnapshotKind
  include_memory: boolean
  /** 创建快照时虚拟机的状态。 */
  vm_status?: string
  size_bytes: number
  status: SnapshotStatus
  created_at: string

  has_children: boolean
  is_current: boolean

  /**
   * 是否可删除 / 可恢复。**由后端算好**。
   *
   * 界面不自己按 has_children / is_current 拼判断：那样「哪些条件下能删」
   * 这条规则会存在两份，而界面上的那份迟早与后端不一致——表现为按钮可点、
   * 但点下去必然失败。
   */
  can_delete: boolean
  can_restore: boolean
}

export interface SnapshotList {
  items: Snapshot[]
  /** 上限与已用数量一起返回，界面才能显示「已用/上限」。 */
  quota: number
  used: number
}

export interface CreateSnapshotInput {
  name: string
  description?: string
  /** 是否保存运行现场（含内存）。含内存的快照体积可能数倍于磁盘。 */
  include_memory: boolean
}

export const snapshotApi = {
  list: (vmID: number) => get<SnapshotList>(`/api/v1/vms/${vmID}/snapshots`),

  create: (vmID: number, input: CreateSnapshotInput) =>
    post<TaskRef>(`/api/v1/vms/${vmID}/snapshots`, input),

  restore: (vmID: number, id: number) =>
    post<TaskRef>(`/api/v1/vms/${vmID}/snapshots/${id}/restore`),

  remove: (vmID: number, id: number) => del<TaskRef>(`/api/v1/vms/${vmID}/snapshots/${id}`),

  /**
   * 删除全部快照。
   *
   * 它是一个**整体操作**而不是循环调用单条删除：逐个入队会让"删到第三个
   * 失败"留下一个既删了一部分、又没有地方可跟踪的中间状态。执行器按依赖
   * 顺序（先删后建的）逐个删除。
   *
   * 需要二次验证（与恢复快照同一档）：丢一个还原点与丢掉整条时间线不是
   * 同一件事。
   */
  deleteAll: (vmID: number, skipCurrent = false) =>
    post<{ task_id: number; total: number; skipped: number }>(
      `/api/v1/vms/${vmID}/snapshots/delete-all`,
      { skip_current: skipCurrent },
    ),
}

/** 修复 UEFI 启动项（恢复快照之后最常见的后遗症）。 */
export const nvramApi = {
  repair: (vmID: number) => post<TaskRef>(`/api/v1/vms/${vmID}/nvram/repair`),
}

/** 快照状态的中文名。 */
export const SNAPSHOT_STATUS_LABEL: Record<SnapshotStatus, string> = {
  creating: '创建中',
  ready: '就绪',
  restoring: '恢复中',
  deleting: '删除中',
  error: '失败',
}

/** 快照状态对应的配色。 */
export const SNAPSHOT_STATUS_TONE: Record<SnapshotStatus, 'success' | 'idle' | 'warning' | 'danger'> =
  {
    creating: 'idle',
    ready: 'success',
    // 恢复中用 warning 而不是 danger：它是一个**正在进行**的动作，
    // 不是出了问题。
    restoring: 'warning',
    deleting: 'idle',
    error: 'danger',
  }

/** 快照种类的中文名。 */
export const SNAPSHOT_KIND_LABEL: Record<SnapshotKind, string> = {
  internal: '内部快照',
  external: '外部快照',
}

/**
 * 快照种类说明。
 *
 * 写"代价是什么"而不是"技术上是什么"：外部快照不保存内存，用户点"创建"
 * 时如果没注意到这一点，会在恢复后发现运行现场不在——那种落差无法挽回。
 */
export const SNAPSHOT_KIND_HINT: Record<SnapshotKind, string> = {
  internal: '保存在同一个镜像文件里，可保存运行现场（内存）。',
  external: '把当前状态另存一份，不含内存——运行中创建的快照属于这一类。',
}

File: internal/model/snapshot.go
package model

import "time"

// 快照种类。
const (
	// SnapshotKindInternal 保存在磁盘镜像内部。
	//
	// 含内存时可恢复**运行现场**，代价是快照体积可能数倍于磁盘本身，
	// 且对磁盘 I/O 有持续影响。
	SnapshotKindInternal = "internal"
	// SnapshotKindExternal 以 `--disk-only` 方式另存为独立文件，只能恢复磁盘。
	SnapshotKindExternal = "external"
)

// 快照状态。
//
// 创建与恢复都是异步任务，界面需要在这期间展示中间状态——这正是快照必须
// 落库的原因之一：虚拟化层不提供「正在创建」这种状态给我们查。
const (
	SnapshotCreating  = "creating"
	SnapshotReady     = "ready"
	SnapshotRestoring = "restoring"
	SnapshotDeleting  = "deleting"
	SnapshotError     = "error"
)

// VMSnapshot 对应 vm_snapshot 表：一台虚拟机的快照。
//
// 它与 vm 表同样是**投影 + 元数据**的混合体：虚拟化层是权威，本表缓存它
// 并补充控制面自己的信息（归属、描述、任务状态）。因此 status 显示为 ready
// 只代表**上次对账时**它是就绪的。
type VMSnapshot struct {
	ID     int64 `gorm:"primaryKey"`
	VMID   int64 `gorm:"not null;uniqueIndex:uniq_vm_snapshot_vm_name,priority:1;index:idx_vm_snapshot_vm_created,priority:1"`
	NodeID int64 `gorm:"not null"`

	Name        string  `gorm:"size:128;not null;uniqueIndex:uniq_vm_snapshot_vm_name,priority:2"`
	Description *string `gorm:"size:255"`

	Kind          string `gorm:"size:16;not null"`
	IncludeMemory bool   `gorm:"not null;default:false"`

	// VMStatus 是创建快照那一刻虚拟机所处的状态。
	//
	// 恢复时据此判断是否必须先关机：把运行中虚拟机的磁盘状态直接回滚，
	// 得到的是一个**文件系统可能已损坏**的来宾系统——它也许还能启动，
	// 然后在某个随机时刻崩溃，而那时没有任何线索指向这次恢复。
	VMStatus *string `gorm:"size:16"`

	// DomainName 是虚拟化层里的快照标识。
	//
	// 与 Name 分开记录：Name 是用户可见、可改的名字；DomainName 一旦建立
	// 就不该再变，变了就找不到那条快照。把两者合成一个字段，会让「重命名
	// 快照」这个很自然的需求变成一次危险操作。
	DomainName *string `gorm:"size:255"`

	SizeBytes int64 `gorm:"not null;default:0"`

	// ParentID 指向父快照。外部快照会形成链，链上的中间节点被删除会让
	// 子快照失去依赖——因此删除前要检查 HasChildren。
	ParentID *int64

	Status string `gorm:"size:16;not null;default:creating"`

	// HasChildren 表示是否存在子快照。界面据此禁用删除并说明原因，
	// 而不是让用户点下去才收到一个拒绝。
	HasChildren bool `gorm:"not null;default:false"`

	// IsCurrent 表示虚拟机当前是否运行在这个快照上。
	// 为 true 时恢复它没有意义，界面应把「恢复」置灰。
	IsCurrent bool `gorm:"not null;default:false"`

	CreatedBy *int64
	CreatedAt time.Time `gorm:"index:idx_vm_snapshot_vm_created,priority:2"`
	UpdatedAt time.Time
}

// TableName 固定表名。
func (VMSnapshot) TableName() string { return "vm_snapshot" }

// IsReady 报告快照是否可用于恢复。
//
// 只有 ready 能恢复：creating 时快照本身还没写完，error 时它不可信。
func (s *VMSnapshot) IsReady() bool {
	return s.Status == SnapshotReady
}

// CanDelete 报告快照是否可以被删除。
//
// 两个条件缺一不可：**没有子快照**（否则会破坏依赖链）、**不是当前快照**
// （删掉它意味着虚拟机正运行在一个已不存在的快照上）。
func (s *VMSnapshot) CanDelete() bool {
	return s.Status != SnapshotCreating && s.Status != SnapshotDeleting &&
		!s.HasChildren && !s.IsCurrent
}
